shapeworks_cloud/core/rest.py

```diff
@@ -302,6 +304,105 @@ def download(self, request, **kwargs):
         project = self.get_object()
         return Response(serializers.ProjectDownloadSerializer(project).data)
 
+    @action(
+        detail=True,
+        url_path='landmarks',
+        url_name='landmarks',
+        methods=['POST'],
+    )
+    def set_landmarks(self, request, **kwargs):
+        project = self.get_object()
+        form_data = request.data
+        landmarks_info = form_data.get('info')
+        landmarks_locations = form_data.get('locations')
+
+        project.landmarks_info = landmarks_info
+        project_file_contents = json.loads(project.file.read())
+        project_file_contents['landmarks'] = landmarks_info
+        project.file.save(
+            project.file.name.split('/')[-1],
+            ContentFile(json.dumps(project_file_contents).encode()),
+        )
+        project.save()
+
+        ids_existing_with_coords = []
+        for subject_id, data in landmarks_locations.items():
+            target_subject = models.Subject.objects.get(id=subject_id)
+            for anatomy_type, locations in data.items():
+                landmarks_object, created = models.Landmarks.objects.get_or_create(
+                    project=project, subject=target_subject, anatomy_type=anatomy_type
+                )
+                file_content = ''
+                if (
+                    locations is not None
+                    and len(locations) > 0
+                    and locations[0] is not None
+                    and len(locations[0]) == 3
```

Comment:
Could you clarify why these are needed?

Reply:
This if statement is intended to ensure that incoming location data is valid and prevent any server error in the case of invalid format. The list of landmark locations must be a list with at least one element, and each element should be an [x, y, z] coordinate.